### ISSUE TITLE
Bug: [no-floating-promises] Incorrect handling of finally calls

### ISSUE THREAD

**Issue by aleclarson** [typescript-eslint/typescript-eslint#5692]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.3&sourceType=module&code=MYewdgzgLgBAJgUwLYgGIFczCgS3ALhgAocwAHdKQgIxBABsEBDMAShgF4A+GABQCcQSHBAQAeaP1IBzHh2KkKVGLQbM2nHgG8AUDBg4AZgvKV2u-fv4Io6fmD6DhogHTWIDAG4IiAcgAWCPT0IDAA7iD89HC+rADcejAAvjBBojAWlta29o5CIghuCABWCNh+ipThTBAwhkz0orEJ+kk6bTqIKBhYuOBEUPzoCKwuUIFgRETs3BmJoJBqLiHSfgCiYFAI1nAw4whgze2jhqQN9ACeUzPa8+AejMsgq74bWzt1Z8EXR0nxQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
demoFunction(true).then(() => {
  console.log('Entered then');
}).finally(() => {
  console.log('Entered finally');
});
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-floating-promises": ["error"],
  },
};
```

### Expected Result

There is no `.catch` call, and the rule should report it.

### Actual Result

Nothing reported

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.31.0` |
| `@typescript-eslint/parser`            | `5.31.0` |
| `TypeScript`                           | `4.7.4` |
| `ESLint`                               | `8.20.0` |
| `node`                                 | `18.7.0` |

**Comment by JoshuaKGoldberg** [comment]
This is intentional. `.finally()` creates a new Promise, and the code inside it might throw an error.

```ts
console.log("a");
Promise.resolve().finally(() => {
  throw new Error("Finally!");
});
console.log("b");
```

```plaintext
a
b
Uncaught (in promise) Error: Finally!
    at <anonymous>:3:9
    at <anonymous>
```

This question comes up once in a while. Let's add an FAQ section to the rule's page explaining why `.finally()` calls are treated as new Promises. Thanks for bringing it up!

**Comment by JoshuaKGoldberg** [comment]
Hmm, we'd explicitly made this the behavior in #1603 -> #1620. But now I'm not so sure this was the right move. `.finally()` creates a new Promise, and the code inside it might throw an error.

```ts
console.log("a");
Promise.resolve().finally(() => {
  throw new Error("Finally!");
});
console.log("b");
```

```plaintext
a
b
Uncaught (in promise) Error: Finally!
    at <anonymous>:3:9
    at <anonymous>
```

cc @a-tarasyuk @bradzacher for historical context. Perhaps we should make the ignoring of `.finally(...)` calls behind an option for the rule?

**Comment by aleclarson** [comment]
Floating promises can't be entirely avoided, since `.catch` handlers can throw. That said, the real issue here is that `.finally` handlers are really just sugar for `.then(handle, handle)` and so they should be treated identically in the `no-floating-promises` rule.

**Comment by aleclarson** [comment]
[Another case][1] that should probably be fixed:
```ts
// Nothing reported by linter.
demoFunction(true).then(() => {
  console.log('Entered then');
}, (error) => {
  console.log('Entered catch');
});
```

In this case (perhaps unintuitively), errors thrown in the fulfill handler **aren't caught** by the rejection handler.

The above code is equivalent to [this monstrosity][2]…

```ts
const catchReturn = Symbol.for('promise.catchReturn')
demoFunction(true).catch((error) => {
  console.log('Entered catch');
  throw { [catchReturn]: undefined }
}).then(() => {
  console.log('Entered then');
}).catch((error) => {
  if (error && error[catchReturn]) {
    return error[catchReturn];
  }
  throw error;
});
```

…which *should be* (but isn't) reported by the `no-floating-promises` rule. The reason I say it *should* be reported is that the last `.catch` handler contains a `throw` statement, which isn't handled anywhere.

We *can* get an equivalent implementation of my last two examples to be reported by the `no-floating-promises` rule by changing it to be [the following][3]…

```ts
const caught = Symbol.for('promise.caught')
demoFunction(true).catch(() => {
  console.log('Entered catch');
  return caught
}).then((result) => {
  if (result !== caught) {
    console.log('Entered then');
  }
});
```

…so I think it stands to reason that the first 2 examples should also be reported.

[1]: https://typescript-eslint.io/play/#ts=4.8.3&sourceType=module&code=MYewdgzgLgBAJgUwLYgGIFczCgS3ALhgAocwAHdKQgIxBABsEBDMAShgF4A+GABQCcQSHBAQAeaP1IBzHh2KkKVGLQbM2nHgG8AUDBg4AZgvKV2u-fv4Io6fmD6DhogHTWIDAG4IiAcgAWCPT0IDAA7iD89HC+rADcejAAvjBBojAWlta29o5CIghuCABWCNh+ipThTBAwhkz0orEJ+kk6bTqIKBhYuOBEUPzoCKwuUIFgRETs3BmJoJBqLiHSfgCiYFAI1nAw4whgze0ANMTbgvwz2vPgHozLIKu+G1s7MMBMUMD+R0nxQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA
[2]: https://typescript-eslint.io/play/#ts=4.8.3&sourceType=module&code=MYewdgzgLgBAJgUwLYgGIFczCgS3ALhgAocwAHdKQgIxBABsEBDMAShgF4A+GABQCcQSHBAQAeaP1IBzHh2KkKVGLQbM2nHgG8AUDBg4AZgvKV2u-fv4Io6fmD6DhogHTWIDAG4IiAcgAWCPT0IDAA7iD89HC+rADcejAAvjBBojAWlta29o5CIghuCABWCNh+ipThTBAwhkz0orEJ+kk6bTqgkLDATFDA-gBKNnYO8gDKAJ5ItPQuhpF+ZE4FLr39QyP2sTqIKBhYuOBEUPzoCKxrfQNERAj8gvzs3BmJXR6MLiHSfgCiYFB7gg4DB1gNmokoP5BGEMjAANpgzY5MAAXUImEQhlIwOS7UuUIQYFuz20b3AH0K3z+AKBIMJYAhSUuSNu90epNe+iMxHZkRgADIBakHpFEddkaNUeZElYtg4+fxxRthijUS08fooTCRY8Esy4kA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA
[3]: https://typescript-eslint.io/play/#ts=4.8.3&sourceType=module&code=MYewdgzgLgBAJgUwLYgGIFczCgS3ALhgAocwAHdKQgIxBABsEBDMAShgF4A+GABQCcQSHBAQAeaP1IBzHh2KkKVGLQbM2nHgG8AUDBg4AZgvKV2u-fv4Io6fmD6DhogHTWIDAG4IiAcgAWCPT0IDAA7iD89HC+rADcejAAvjBBojAWlta29o5CIghuCABWCNh+ipThTBAwhkz0orEJ+kk6bTqgkLDATOjS-rDyAMoAnki09C6GkX5kTgUuvf2DsTqIKBhYuOBEUPzoCKxLTFDA-kRE7NwZiV0ejC4h0n4AomBQCNZwML1n-s1Etk7A5lgMoO1jlBAmBLu50PQoNdtIkjMR4YiYABCDjyMGDcyJfT3NRPEAvXzvT7fGDQhBgQGtSFxIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

**Comment by bradzacher** [comment]
In #1620 it was written that `.finally` should be treated like a `.catch`, which isn't correct, because `.finally` is "transparent" in regards to the promise chain - the value returned from `.finally` does not change the settled value of the promise unless the `.finally` handler throws an error.

So what should the handling be?

I think that `.finally` should essentially be completely ignored by the rule - it shouldn't end a chain, it shouldn't start a new chain - it should just be skipped over and the rule should pretend like it wasn't there.

I.e.
```ts
// BAD
prom;
prom.then(f); 
prom.finally(f);
prom.then(f).finally(f);
prom.finally(f).then(f).finally(f);

// GOOD
prom.then(f, f).finally(f);
prom.catch(f).finally(f);
prom.then(f).catch(f).finally(f);
prom.then(f).finally(f).catch(f).finally(f);
prom.catch(f).finally(f).finally(f).finally(f).finally(f).finally(f).finally(f).finally(f);
```

**Comment by kirkwaiblinger** [comment]
I am working on a PR (https://github.com/typescript-eslint/typescript-eslint/pull/6881) which forced me to think about this a bit... For the purposes of that PR I regarded `.finally` as _different_ from the rejection handlers in `.then` and `.catch`, though I noticed that the existing code treated them the same.

I'd be happy to look into putting up a PR for this issue as well.

To build off of @bradzacher's suggestions:

I also noticed that from the point of view of avoiding literal unhandled rejections, finally can only make an unhandled promise worse. However, the more I think about it, I can also understand the argument that adding a `.finally` means you are, well, handling the rejection. Not in the sense of preventing a rejection from bubbling up, but in the sense of you, the developer, are acknowledging a rejection could occur, and taking some mitigating action. It's similar to being able to mark floating promises as ignorable with the 'void' operator - you know that they can reject, but you're not worried about it.

My preference for this rule would be to make a `.finally` totally transparent, regardless of whether it has a handler or not. 
```
asyncFunction().finally() // Currently bad, Propose: still bad
asyncFunction().finally(f) // Currently good, Propose: bad
asyncFuncion().catch(c).finally() // Currently bad, Propose: good
asyncFunction().catch(c).finally(f) // Currently good, Propose: still good
```

JS/TS is not an exception-safe language, so I think it would be overbearing to consider the possibility of the finally handler throwing. Actually, I think that a `.catch` handler is possibly more likely to throw than a `.finally` handler in practice, since:
```
// This is common in synchronous code.
try {
    mightThrow()
} catch (e) {
    if (isExpectedError(e)) {
        // handle error
    }
    throw error;
}

// It's not crazy to me that this would be present in async code
mightReject().catch(e => {
   if (isExpectedError(e)) {
       // handle
   } 
   throw e;
})
```

So let's just take the fact that `.finally` is transparent to the promise status at face value.

One more thought, just to take this argument ad absurdum, by turning to synchronous equivalent again:
```
try {
    mightThrow()
} catch (e) { 
   // handle error
} finally {
   // have you ever seen anyone put exception-safe code here ???
   try {
       // do the cleanup work
   } catch (e) {
       // handle finally handler error 
       return;
   }
}
```
finally handlers are never truly safe - that's just a limitation of JS. The linter can't compensate for this.

**Comment by aleclarson** [comment]
My attempt to reduce the desired behavior (IMO) to its clearest and simplest form is as follows:

- Any `.then` call that is neither returned nor awaited **must** have a `.catch` call somewhere after it for the linter to be satisfied

- Ending a promise chain with `.then(onResult, onError)` should **not** satisfy the linter, because errors in `onResult` won't be caught by the `onError` function

- Ending a promise chain with `.then(onResult).finally(onResolve)` should **not** satisfy the linter, because a rejected promise from `.then` will pass through `.finally` and still be an unhandled rejection

- Possibility of an uncaught error in a `.catch` handler, a `.finally` handler, or the `onError` argument of `.then(onResult, onError)` are considered an unavoidable hazard (therefore, the linter should ignore them)

Does that clear things up?
